# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsGlobalWindowOuter.cpp
# Commit: d7e962f7ca16
# Full Hash: d7e962f7ca165897bc5c1e32f8a1bcc72ab24629
# Author: Jonathan Watt <jwatt@jwatt.org>
# Date: 2020-01-22 21:48:55
# Regressor Bug: 1602410
# File Overlap Count: 3
# Description:
#   Bug 1607671. Back out changeset edf3df2adebb (bug 1602410) for regressions. kmag r=RyanVM
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D59785
# ==============================================================================

diff -r 61c5e64eb2b4 -r d7e962f7ca16 dom/base/nsGlobalWindowOuter.cpp
--- a/dom/base/nsGlobalWindowOuter.cpp	Wed Jan 22 14:41:47 2020 +0000
+++ b/dom/base/nsGlobalWindowOuter.cpp	Wed Jan 22 14:43:46 2020 +0000
@@ -25,7 +25,6 @@
 #include "mozilla/dom/BrowserChild.h"
 #include "mozilla/dom/ContentFrameMessageManager.h"
 #include "mozilla/dom/EventTarget.h"
-#include "mozilla/dom/JSWindowActorChild.h"
 #include "mozilla/dom/LocalStorage.h"
 #include "mozilla/dom/LSObject.h"
 #include "mozilla/dom/Storage.h"
@@ -4973,59 +4972,69 @@
     return;
   }
 
-  nsCOMPtr<nsIPrintSettingsService> printSettingsService =
-      do_GetService("@mozilla.org/gfx/printsettings-service;1");
-  if (!printSettingsService) {
-    aError.Throw(NS_ERROR_NOT_AVAILABLE);
-    return;
-  }
-
-  WindowGlobalChild* wgc = mInnerWindow->GetWindowGlobalChild();
-  if (!wgc) {
-    aError.Throw(NS_ERROR_NOT_AVAILABLE);
-    return;
-  }
-
-  RefPtr<JSWindowActorChild> actor =
-      wgc->GetActor(NS_LITERAL_STRING("BrowserElement"), IgnoreErrors());
-  if (!actor) {
-    aError.Throw(NS_ERROR_NOT_AVAILABLE);
-    return;
-  }
-
-  bool havePrintableSelection = false;
-  if (!mDoc->GetRootElement()->HasAttr(kNameSpaceID_None,
-                                       nsGkAtoms::mozdisallowselectionprint)) {
-    if (Selection* selection = GetSelectionOuter()) {
-      if (int32_t rangeCount = selection->RangeCount()) {
-        havePrintableSelection =
-            rangeCount > 1 ||
-            // check to make sure it isn't an insertion selection
-            (selection->GetRangeAt(0) && !selection->IsCollapsed());
+  nsCOMPtr<nsIWebBrowserPrint> webBrowserPrint;
+  if (NS_SUCCEEDED(GetInterface(NS_GET_IID(nsIWebBrowserPrint),
+                                getter_AddRefs(webBrowserPrint)))) {
+    nsAutoSyncOperation sync(GetCurrentInnerWindowInternal()
+                                 ? GetCurrentInnerWindowInternal()->mDoc.get()
+                                 : nullptr);
+
+    nsCOMPtr<nsIPrintSettingsService> printSettingsService =
+        do_GetService("@mozilla.org/gfx/printsettings-service;1");
+
+    nsCOMPtr<nsIPrintSettings> printSettings;
+    if (printSettingsService) {
+      bool printSettingsAreGlobal =
+          Preferences::GetBool("print.use_global_printsettings", false);
+
+      if (printSettingsAreGlobal) {
+        printSettingsService->GetGlobalPrintSettings(
+            getter_AddRefs(printSettings));
+
+        nsAutoString printerName;
+        printSettings->GetPrinterName(printerName);
+
+        bool shouldGetDefaultPrinterName = printerName.IsEmpty();
+#  ifdef MOZ_X11
+        // In Linux, GTK backend does not support per printer settings.
+        // Calling GetDefaultPrinterName causes a sandbox violation (see Bug
+        // 1329216). The printer name is not needed anywhere else on Linux
+        // before it gets to the parent. In the parent, we will then query the
+        // default printer name if no name is set. Unless we are in the parent,
+        // we will skip this part.
+        if (!XRE_IsParentProcess()) {
+          shouldGetDefaultPrinterName = false;
+        }
+#  endif
+        if (shouldGetDefaultPrinterName) {
+          printSettingsService->GetDefaultPrinterName(printerName);
+          printSettings->SetPrinterName(printerName);
+        }
+        printSettingsService->InitPrintSettingsFromPrinter(printerName,
+                                                           printSettings);
+        printSettingsService->InitPrintSettingsFromPrefs(
+            printSettings, true, nsIPrintSettings::kInitSaveAll);
+      } else {
+        printSettingsService->GetNewPrintSettings(
+            getter_AddRefs(printSettings));
       }
-    }
-  }
-
-  // We don't init the AutoJSAPI with ourselves because we don't want it
-  // reporting errors to our onerror handlers.
-  AutoJSAPI jsapi;
-  jsapi.Init();
-  JSContext* cx = jsapi.cx();
-  JSAutoRealm ar(cx, actor->GetParentObject()->GetGlobalJSObject());
-
-  // The frontend code doesn't yet make use of the `data` we send here, but it
-  // will do in future work in order to avoid a roundtrip to the content
-  // process.
-  // Note: if we want more complex detail, create something like
-  // DOMWindowResizeEventDetail.
-  JS::Rooted<JS::Value> msgData(cx, JS::BooleanValue(havePrintableSelection));
-
-  IgnoredErrorResult res;
-  actor->SendAsyncMessage(cx, NS_LITERAL_STRING("ContentRequestedPrint"),
-                          msgData, res);
-  if (res.Failed()) {
-    aError.Throw(NS_ERROR_NOT_AVAILABLE);
-    return;
+
+      EnterModalState();
+      webBrowserPrint->Print(printSettings, nullptr);
+      LeaveModalState();
+
+      bool savePrintSettings =
+          Preferences::GetBool("print.save_print_settings", false);
+      if (printSettingsAreGlobal && savePrintSettings) {
+        printSettingsService->SavePrintSettingsToPrefs(
+            printSettings, true, nsIPrintSettings::kInitSaveAll);
+        printSettingsService->SavePrintSettingsToPrefs(
+            printSettings, false, nsIPrintSettings::kInitSavePrinterName);
+      }
+    } else {
+      webBrowserPrint->GetGlobalPrintSettings(getter_AddRefs(printSettings));
+      webBrowserPrint->Print(printSettings, nullptr);
+    }
   }
 #endif  // NS_PRINTING
 }